radicalxdev/marvel-platform
# Implemented the V1 notifications system with the corresponding unit t…

## Description

This PR is for the implementation of Epic 1.5, the V1 Notification System

## Related Issue

If this PR addresses an issue, link to it here.
https://github.com/marvelai-org/marvel-platform/issues/201

## Type of Change

Please select the type(s) of change that apply and delete those that do not. 


## Proposed Solution

### New dependencies:

- @babel/preset-env: Babel preset for test cases
- @babel/preset-react: Babel preset for react test cases.
- @firebase/testing: Testing library for the backend trigger unit tests.
- @testing-library/dom: Required dependency of @testing-library/react used to unit test dom components.
- @testing-library/react: Used to test react components.
- babel-jest: Allows babel configuration for jest test cases of react components
- cross-fetch: Enables fetch for backend trigger test cases
- firestore-jest-mock: Allows mocking of firestore in jest.
- jest: Testing library used for test cases
- jest-environment-jsdom: Allows the jsdom test environment for react unit tests.
- Reactfire: Used in place of websockets to retrieve notifications in real time from firestore.

### Updated Dependency:
- firebase-admin@11.8.0 -> firebase-admin@11.9.0: Fixes a bug where Filter was not added to firebase-admin/firestore. Used in backend trigger unit tests.

### Implementation reasonings

- Real-time notifications was implemented with reactfire. This is a library that can retrieve updates to firestore in real-time. This library was used because it is much easier to set up and reuse than websockets.
- The backend has two implemented collections 'notifications' and 'personal-notifications'. 'notifications' are meant for global notifications passed to every user like tool updates. 'personal-notifications' are meant to be notifications that are unique to users like personal alerts. 

### NotificationsButton.jsx (New)

This component displays a bell icon. On click this toggles the visibility of a list of Notifications.

### NotificationList.jsx (New)

This component displays: A togglable all/unread button. A list of notifications relevant to the user. A button to read all notifications. This component is designed to be reused to display a list of notifications anywhere.

### Notification.jsx (New)

This component displays: A single notification.

### Firebase.jsx and FirebaseProviders.jsx (New)

These components are wrappers that configure reactfire to retrieve data from the configured firebase destination.

### setReadStatus.js (New)

This function takes a message and updates the existing message in the database.

### SideMenu.jsx (Updated)

Added the notification button to this Layout.

### MainAppLayout/styles.js (Updated)

Updated the zIndex in the navigation bar, so the notifications appear above the contents of the page.

### notificationController.js (New)

- getNotifications: Takes a user id and an optional notification id and queries the personal-notifications collection. Returns a list of notifications that matches the query.

- setNotificationStats: Takes a user id, an optional notification id, and a status. Queries the personal-notifications collection using user id and notification id and updates the is_read status of the query results with status.

- onNewNotification: Trigger that occurs when a global notification is created. Creates a copy of the notification as a personal notification for every user.

- onRemovedNotification: trigger that occurs when a global notification is removed. Removes every personal notification that was a copy of the global notification.

- onNewTool: Trigger that occurs when a document is added to 'tools'. Creates a global notification that a new tool has been created.

- onUpdatedTool: Trigger that occurs when a document is updated in 'tools'. Creates a global notification that a tool has been updated.

- onNewAssistant: Trigger that occurs when a document is added to 'assistants'. Creates a global notification that a new assistant has been created.

- onUpdatedAssistant: Trigger that occurs when a document is updated in 'assistants'. Creates a global notification that a assistant has been updated.

### cloud_db_seed.js

- Added a global notification and two personal notifications to the seed database.

## How to Test

[Notification System Demo Link on Loom](https://www.loom.com/share/64b0b4885a8849bc83e24c06a29ab533)
- Click the notification button to make sure the button toggles the visibility of the notifications menu when clicked.
- Click a 'Mark as Read' button to make sure it changes to 'Mark as Unread' and vice versa
- Navigate to the Unread tab and mark a notification as read to make sure it disappears from the tab
- Click the 'Mark all as Read' button to make sure all notifications change to read.
- Go to firebase emulators to add a new notification. The website should update with the notification without needing to refresh.

## Unit Tests

Unit Tests are stored in \_\_tests\_\_. Emulator must be running in order for backend trigger tests to work. The backend trigger tests may not work sometimes due to delays due to unresponsiveness from the server.

### \_\_setup\_\_/env-setup.js

- Sets up jest environment for each test by passing .env file and activating cross-fetch for backend triggers.

### jest.config.js

- Configures the jest environment.

### babel.config.js

- Configures jest environment with babel presets.

### NotificationsButton-test.js

- Tests whether the NotificationsButton component starts with the notification list hidden.
- Tests whether clicking the NotificationsButton component toggles the visibility of the notification list.

### NotificationList-test.js

- Tests whether the NotificationsList component displays notifications correctly based on passed information.
- Tests whether the correct information is passed to setReadStatus when 'Mark As Read' is clicked.
- Tests whether the correct information is passed to setReadStatus when 'Mark all as read' is clicked.

### NotificationTriggers-test.js

- Tests whether a personal notification is created for a user when a global notification is created.
- Tests whether a personal notification for a user is removed when the corresponding global notification is removed.
- Tests whether a global notification is created when a new tool is created.
- Tests whether a global notification is created when a tool is updated.
- Tests whether a global notification is created when a new assistant is created.
- Tests whether a global notification is created when an assistant is updated.

## Documentation Updates

Indicate whether documentation needs to be updated due to this PR. 


If yes, describe what documentation updates are needed and link to the relevant documentation.

## Checklist


## Additional Information

 Add any other information that might be useful for the reviewers.
